### ISSUE TITLE
VICReg loss ?

### ISSUE THREAD

**Issue by ralgayres** [KevinMusgrave/pytorch-metric-learning#372]
Hi,
will there be an update with VICReg loss function? It is a pretty cool loss that does not need negative samples.
https://arxiv.org/abs/2105.04906

**Comment by KevinMusgrave** [comment]
Thanks for the suggestion

**Comment by KevinMusgrave** [comment]
> ## Option 1: Wait for other branch to merge

That branch is now merged

**Comment by codeandproduce** [comment]
@KevinMusgrave Right, the implementation seems pretty intuitive with the new merge. Happy to submit a pull request if nobody's working on it right now.

**Comment by KevinMusgrave** [comment]
@codeandproduce Yes please do!

**Comment by codeandproduce** [comment]
@KevinMusgrave 

The function itself is easily done, but I'm not sure if this can fit at all into the current `BaseMetricLossFunction` framework.
```python

class VICRegLoss(BaseMetricLossFunction):
    def __init__(
        self,
        invariance_lambda=25,
        variance_mu=25,
        covariance_v=1,
        eps=1e-4,
        **kwargs
    ):
        super().__init__(**kwargs)
        """
        The overall loss function is a weighted average of the invariance, variance and covariance terms:
            L(Z, Z') = λs(Z, Z') + µ[v(Z) + v(Z')] + ν[c(Z) + c(Z')],
        where λ, µ and ν are hyper-parameters controlling the importance of each term in the loss.
        """
        self.invariance_lambda = invariance_lambda
        self.variance_mu = variance_mu
        self.covariance_v = covariance_v
        self.eps = eps

    def compute_loss(
        self, embeddings, ref_emb
    ):

        invariance_loss = self.invariance_loss(embeddings, ref_emb)
        variance_loss = self.variance_loss(embeddings, ref_emb)
        covariance_loss = self.covariance_loss(embeddings, ref_emb)

        loss = self.invariance_lambda * invariance_loss + self.variance_mu * variance_loss + self.covariance_v * covariance_loss
        return loss

# ...implementation of the above individual loss functions
```

The idea is that if you have a batch of images `I` and some augmented version of those images `I'` (e.g. a flip) and your model embeds these images `E(I) = Z`,

```python
loss_fn = VICRegLoss()

Z = E(I)
Z' = E(I')

loss = loss_fn(Z, Z')
```

It doesn't use any notion of `labels` or `indices_tuple`. 

I was considering forcing this framework with something like this:
```python
# assume Z.size() == (5, 255)
embeds = Z
ref_embeds = Z'
labels = [1,2,3,4,5]

loss = loss_fn(embeds, ref_embeds, labels, ref_labels=labels)
```

but this is very unnecessary.. it's not like this function could benefit from handling cases where `ref_labels` is not exactly the same as `labels`. 

But the specific thing making my current implementation impossible is the following `forward()` function in the `BaseMetricLossFunction` class:

```python
loss_dict = self.compute_loss(
    embeddings, labels, indices_tuple, ref_emb, ref_labels
)
self.add_embedding_regularization_to_loss_dict(loss_dict, embeddings)
return self.reducer(loss_dict, embeddings, labels)
```

since it expects a dict that looks something like:

```python
 return {
    "loss": {
        "losses": loss,
        "indices": indices_tuple,
        "reduction_type": "triplet",
    }
}
```

and `VICRegLoss` doesn't use any notion of `indices` or `reduction`. 

I'm wondering if I should just write a new Base class with something like `BaseSimpleLossFunction` that just directly returns the loss of the `compute_loss` function of the child class with the purpose of being the most bare-bones and flexible superclass.

```python
class BaseSimpleLossFunction:
    def compute_loss(self, embeddings, labels, indices_tuple=None):
        """
        This has to be implemented and is what actually computes the loss.
        """
        raise NotImplementedError

    def forward(
        self, embeddings=None, labels=None, indices_tuple=None, ref_emb=None, ref_labels=None
    ):
        loss = self.compute_loss(
            embeddings, labels, indices_tuple, ref_emb, ref_labels
        )
        return loss

    def zero_loss(self):
        return {"losses": 0, "indices": None, "reduction_type": "already_reduced"}

    def zero_losses(self):
        return {loss_name: self.zero_loss() for loss_name in self.sub_loss_names()}

    def _sub_loss_names(self):
        return ["loss"]

    def sub_loss_names(self):
        return self._sub_loss_names() + self.all_regularization_loss_names()

```

**Comment by KevinMusgrave** [comment]
I guess ```BaseRegularizer``` is already kind of like your ```BaseSimpleLossFunction```. But it is missing a ```ref``` argument from its forward function. 

For now, you could try overriding the forward function and see if the design is reasonable:

```python
from .base_regularizer import BaseRegularizer
from ..utils import common_functions as c_f

class VICRegLoss(BaseRegularizer):
    def forward(self, x, y):
        self.reset_stats()
        loss_dict = self.compute_loss(x)
        return self.reducer(loss_dict, x, c_f.torch_arange_from_size(x))

    def compute_loss(self, x, y):
        invariance_loss = self.invariance_loss(embeddings, ref_emb)
        variance_loss = self.variance_loss(embeddings, ref_emb)
        covariance_loss = self.covariance_loss(embeddings, ref_emb)
        loss = self.invariance_lambda * invariance_loss + self.variance_mu * variance_loss + self.covariance_v * covariance_loss
        return {
            "loss": {
                "losses": loss,
                "indices": None,
                "reduction_type": "already_reduced",
            }
        }
```

It might be possible to keep the losses unreduced. For example, the mean square distance between vectors can be kept as per-pair (i.e. not passed into torch.mean).

The "losses", "indices", and "reduction_type" keys are a bit confusing. I think the reducers are overly complicated in general. Anyway, if you decide to keep the losses unreduced, you can return something like this:

```python
        return {
            "invariance_loss": {
                "losses": invariance_loss,
                "indices": c_f.torch_arange_from_size(),
                "reduction_type": "element",
            }
            # repeat for the other 2 losses. The other two losses can be
        }
```

But to keep things simple, I would start with the "already_reduced" version.

**Comment by KevinMusgrave** [comment]
Re: the components of the VIC loss, you could write them as simple loss functions (i.e. extend ```torch.nn.Module```), and then use them inside of VICRegLoss.

**Comment by KevinMusgrave** [comment]
Ultimately, if you find it makes more sense to extend ```torch.nn.Module```, then that's ok too.

**Comment by KevinMusgrave** [comment]
Available in v1.1:
```
pip install pytorch-metric-learning==1.1.0
```
